work_queue/src/work_queue.c

```diff
@@ -827,6 +855,127 @@ static int get_transfer_wait_time(struct work_queue *q, struct work_queue_worker
 	return timeout;
 }
 
+static int factory_trim_workers(struct work_queue *q, struct work_queue_factory_info *f)
+{
+	// Iterate through all workers and shut idle ones down
+	char *factory_name = f->name;
+	struct work_queue_worker *w;
+	char *key;
+	int trimmed_workers = 0;
+
+	hash_table_firstkey(q->worker_table);
+	while( f->connected_workers > f->max_workers &&
+			hash_table_nextkey(q->worker_table, &key, (void **) &w) ) {
+		if( !strcmp(factory_name, w->factory_name) &&
+			!itable_size(w->current_tasks) ) {
+			shut_down_worker(q, w);
+			trimmed_workers++;
+		}
+	}
+
+	debug(D_WQ, "Trimmed %d workers from %s", trimmed_workers, factory_name);
+	return trimmed_workers;
+}
+
+static struct work_queue_factory_info *create_factory_info(struct work_queue *q, const char *name)
+{
+	struct work_queue_factory_info *f;
+	if ( (f = hash_table_lookup(q->factory_table, name)) ) return f;
+	if ( (f = malloc(sizeof(*f))) ) {
+		f->name = xxstrdup(name);
+		f->connected_workers = 0;
+		f->max_workers = INT_MAX;
+		f->seen_at_catalog = 0;
+	} else {
+		debug(D_WQ, "Failed to allocate memory for factory info");
+	}
+	hash_table_insert(q->factory_table, name, f);
+	return f;
+}
+
+static void remove_factory_info(struct work_queue *q, const char *name)
+{
+	struct work_queue_factory_info *f = hash_table_lookup(q->factory_table, name);
+	if (f) {
+		free(f->name);
+		free(f);
+		hash_table_remove(q->factory_table, name);
+	} else {
+		debug(D_WQ, "Failed to remove unrecorded factory %s", name);
+	}
+}
+
+static void update_factory(struct work_queue *q, struct jx *j)
+{
+	const char *name;
+	if ( !(name = jx_lookup_string(j, "factory_name")) ) return;
```

Comment:
```
name = jx_lookup_string(j, "factory_name");
if(!name) {
   return;
}
```

Reply:
I'm not sure what's the difference between your code and mine?

Reply:
Mine is easier to read and debug. Assignments inside `if`s are sometimes confused with buggy `==`.